azkaban-common/src/main/java/azkaban/trigger/builtin/SlaChecker.java

```diff
@@ -32,210 +32,184 @@
-        ReadablePeriod duration =
-            Utils.parsePeriodString((String) slaOption.getInfo().get(
-                SlaOption.INFO_DURATION));
-        DateTime startTime = new DateTime(node.getStartTime());
-        DateTime nextCheckTime = startTime.plus(duration);
-        this.checkTime = nextCheckTime.getMillis();
-      }
-      status = node.getStatus();
-      return isJobFinished(status);
-    } else if (type.equals(SlaOption.TYPE_JOB_SUCCEED)) {
-      String jobName =
-          (String) slaOption.getInfo().get(SlaOption.INFO_JOB_NAME);
-      ExecutableNode node = flow.getExecutableNode(jobName);
-      if (node.getStartTime() < 0) {
-        return Boolean.FALSE;
+    return null;
+  }
+
+  private boolean init() {
+    try {
+      requireNonNull(executorManager);
+      if (flow == null) {
+        flow = executorManager.getExecutableFlow(execId);
+        if (NodeType.JOB == nodeType) {
+          String jobName = (String) slaOption.getInfo().get(SlaOption.INFO_JOB_NAME);
+          job = flow.getExecutableNode(jobName);
+        }
       }
-      if (checkTime < node.getStartTime()) {
-        ReadablePeriod duration =
-            Utils.parsePeriodString((String) slaOption.getInfo().get(
-                SlaOption.INFO_DURATION));
-        DateTime startTime = new DateTime(node.getStartTime());
-        DateTime nextCheckTime = startTime.plus(duration);
-        this.checkTime = nextCheckTime.getMillis();
+      // Start time needs to be initialized if unset
+      if (startTime < 0) {
+        long t = getStartTime();
+        if (t > 0) {
+          this.startTime = t;
+          ReadablePeriod duration = Utils.parsePeriodString((String) slaOption.getInfo().get(SlaOption.INFO_DURATION));
+          this.slaExpireTime = new DateTime(startTime).plus(duration).getMillis();
+        }
       }
-      status = node.getStatus();
-      return isJobSucceeded(status);
+    } catch (ExecutorManagerException e) {
+      // Log errors but don't crash
+      logger.error("Can't get executable flow.", e);
+      return false;
     }
-    return Boolean.FALSE;
+    return true;
+  }
+
+  private boolean hasSlaExpired() {
+    return slaExpireTime < DateTime.now().getMillis();
   }
 
   // return true to trigger sla action
   @Override
   public Object eval() {
     logger.info("Checking sla for execution " + execId);
-    ExecutableFlow flow;
-    try {
-      flow = executorManager.getExecutableFlow(execId);
-    } catch (ExecutorManagerException e) {
-      logger.error("Can't get executable flow.", e);
-      e.printStackTrace();
-      // something wrong, send out alerts
-      return Boolean.TRUE;
-    }
-    return isSlaMissed(flow);
+    return isSlaFailed();
   }
 
+  /**
+   * TODO Remove convoluted logic. Investigate the need for 2 separate methods for pass and fail test
+   * DO NOT DELETE METHOD. This is used by some weird reflection code using JEXL expression
+   *
+   * @return
+   */
+  @SuppressWarnings("unused")
```

Comment:
yes. there are multiple code paths and am not sure of all of them. The reason I haven't changed the public APIs is because of JEXL. If / when I touch the trigger management code itself, I can refactor the logic and clean up a lot of the complexities. For now, I'm leaving them as is.